cleanup-passes/typescript.ts

```diff
@@ -0,0 +1,150 @@
+import * as path from 'path';
+import {execFile} from 'child_process';
+import * as yaml from 'js-yaml';
+import {promisify} from 'util';
+
+import {register} from '../cleanup-pass';
+import {ElementRepo} from '../element-repo';
+import {makeCommit} from './util';
+
+const execFilePromise = promisify(execFile);
+
+const generatorPackageName = '@polymer/gen-typescript-declarations';
+const generatorSemver = '^1.1.1';
+const npmScriptName = 'update-types';
+const npmScriptCommand =
+    'bower install && gen-typescript-declarations --deleteExisting --outDir .';
+
+/**
+ * This pass updates an element repo to include typings generated by
+ * https://github.com/Polymer/gen-typescript-declarations/ and configures an
+ * "update-types" NPM script that can be run to re-generate them.
+ *
+ * Throws an error if the repo has a missing or invalid package.json or if npm
+ * or bower aren't globally installed.
+ */
+async function typescriptPass(element: ElementRepo): Promise<void> {
+  const packageJsonPath = path.join(element.dir, 'package.json');
+  let packageJson: NpmConfig;
+  try {
+    packageJson = await fse.readJson(packageJsonPath);
+  } catch {
+    throw new Error(`${element.ghRepo.name}: Missing or invalid package.json.`);
+  }
+
+  if (packageJson.devDependencies === undefined) {
+    packageJson.devDependencies = {};
+  }
+  if (packageJson.scripts === undefined) {
+    packageJson.scripts = {};
+  }
+
+  let writePackageJson = false;
+  if (packageJson.devDependencies[generatorPackageName] !== generatorSemver) {
+    packageJson.devDependencies[generatorPackageName] = generatorSemver;
+    writePackageJson = true;
+  }
+  // The update-types script depends on bower because the generator needs to
+  // actually resolve dependencies in the HTML import graph.
+  if (packageJson.devDependencies['bower'] === undefined) {
+    packageJson.devDependencies['bower'] = '^1.8.0';
+    writePackageJson = true;
+  }
+  if (packageJson.scripts[npmScriptName] !== npmScriptCommand) {
+    packageJson.scripts[npmScriptName] = npmScriptCommand;
+    writePackageJson = true;
+  }
+  if (writePackageJson) {
+    await fse.writeJson(packageJsonPath, packageJson, {spaces: 2});
+  }
+
+  // Update Travis config to fail if typings aren't up to date.
+  const travisYamlPath = path.join(element.dir, '.travis.yml');
+  if (await fse.pathExists(travisYamlPath)) {
+    const travisYaml =
+        yaml.safeLoad(await fse.readFile(travisYamlPath, 'utf8')) as {
+      before_script?: string[];
+    }
+    if (!travisYaml.before_script) {
+      travisYaml.before_script = [];
+    }
+    // Remove any prior version of this check.
+    travisYaml.before_script = travisYaml.before_script.filter(
+        (line) => !line.includes('update-types'));
+
+    const travisCommand =
+        // Update the types.
+        'npm run update-types && ' +
+        // If there were any changes, this git command will return non-zero.
+        'git diff --exit-code || ' +
```

Comment:
🎉 🎉 🎉 🎉 

That error echo is really nice. Maybe update the script in polymer/polymer as well? 😄

Reply:
> That error echo is really nice. Maybe update the script in polymer/polymer as well? 😄

Done in https://github.com/Polymer/polymer/pull/5099